README.md
# job-collection

[![Build Status](https://travis-ci.org/SimonSimCity/meteor-job-collection.svg)](https://travis-ci.org/simonsimcity/meteor-job-collection)

## Notice!



#### **This is fork was made to have this package available on meteor 1.6.1 (where we needed to make up with coffeescript 1). It will be maintained by me as I need it for my project. I'm happy to take a look at your contribution if you send me a PR.**

## Intro

job-collection is a powerful and easy to use job manager designed and built for Meteor.js.

It solves the following problems (and more):

* Schedule jobs to run (and repeat) in the future, persisting across server restarts
* Create repeating jobs with complex schedules using [Later.js](https://bunkat.github.io/later/index.html). See demo [here](https://jcplayground.meteorapp.com)
* Move work out of Meteor's single threaded event-loop
* Enable work on computationally expensive jobs to run anywhere, on any number of machines
* Track jobs and their progress, and automatically retry failed jobs
* Easily build an admin UI to manage all of the above using Meteor's reactivity and UI goodness

A live demo of job-collection is at [https://jcplayground.meteorapp.com](https://jcplayground.meteorapp.com).

Source code for this demo app is at [meteor-job-collection-playground](https://github.com/vsivsi/meteor-job-collection-playground)

Source code for a plain node.js worker for the demo app is at [meteor-job-collection-playground-worker](https://github.com/vsivsi/meteor-job-collection-playground-worker)

## What's new in v1.5.0?

* Added `errorCallback` option to `processJobs()`. This gives workers a way to log network and other errors and do something other than write them to `console.error` (the default and previous behavior).
* Progress information is now preserved in the job document on 'done' or 'fail', previously it was set to hard coded values.

A complete list of changes can be found in the HISTORY file.

## Table of Contents

- [Quick example](#user-content-quick-example)
- [Design](#user-content-design)
- [Installation](#user-content-installation)
- [Use](#user-content-use)
  - [Security](#user-content-security)
  - [Performance](#user-content-performance)
  - [Removing Old Jobs](#user-content-removing-old-jobs)
  - [Logging](#user-content-logging)
- [JobCollection API](#user-content-jobcollection-api)
- [Job API](#user-content-job-api)
- [JobQueue API](#user-content-jobqueue-api)
- [Job document data models](#user-content-job-document-data-models)
- [DDP Method reference](#user-content-ddp-method-reference)
- [Zombie jobs](#user-content-zombie-jobs)

### Quick example

The code snippets below show a Meteor server that creates a `JobCollection`, Meteor client code that
subscribes to it and creates a new job, and a pure node.js program that can run *anywhere* and work
on such jobs.

```javascript
///////////////////
// Server
if (Meteor.isServer) {

  var myJobs = new JobCollection('myJobQueue');
  myJobs.allow({
    // Grant full permission to any authenticated user
    admin: function (userId, method, params) {
      return (userId ? true : false);
    }
  });

  Meteor.startup(function () {
    // Normal Meteor publish call, the server always
    // controls what each client can see
    Meteor.publish('allJobs', function () {
      return myJobs.find({});
    });

    // Start the myJobs queue running
    return myJobs.startJobServer();
  });
}
```

Alright, the server is set-up and running, now let's add some client code to create/manage a job.

```javascript
///////////////////
// Client
if (Meteor.isClient) {

  var myJobs = new JobCollection('myJobQueue');

  Meteor.startup(function () {
    Meteor.subscribe('allJobs');

    // Because of the server settings, the code below will
    // only work if the client is authenticated.
    // On the server, all of it would run unconditionally.

    // Create a job:
    var job = new Job(myJobs, 'sendEmail', // type of job
      // Job data that you define, including anything the job
      // needs to complete. May contain links to files, etc...
      {
        address: 'bozo@clowns.com',
        subject: 'Critical rainbow hair shortage',
        message: 'LOL; JK, KThxBye.'
      }
    );

    // Set some properties of the job and then submit it
    job.priority('normal')
      .retry({ retries: 5,
        wait: 15*60*1000 })  // 15 minutes between attempts
      .delay(60*60*1000)     // Wait an hour before first try
      .save();               // Commit it to the server

    // Now that it's saved, this job will appear as a document
    // in the myJobs Collection, and will reactively update as
    // its status changes, etc.

    // Any job document from myJobs can be turned into a Job object
    job = new Job(myJobs, myJobs.findOne({}));

    // Or a job can be fetched from the server by _id
    myJobs.getJob(_id, function (err, job) {
      // If successful, job is a Job object corresponding to _id
      // With a job object, you can remotely control the
      // job's status (subject to server allow/deny rules)
      // Here are some examples:
      job.pause();
      job.cancel();
      job.remove();
      // etc...
    });
  });
}
```

**Q:** Okay, that's cool, but where does the actual work get done?

**A:** Anywhere you want!

job-collection is extremely flexible in where the work can get done; from workers that only run on a
single Meteor server to hundreds of node.js workers running on a cluster or in the cloud. Work can
also be done on other Meteor servers (different from the one hosting the Job collection), or in some
cases it can be done within properly authenticated Meteor clients!

Below is a pure node.js program that can obtain jobs from the server above and "get 'em done".
Powerfully, this can be run ***anywhere*** that has node.js and can connect to the server. The
secret sauce here is the [meteor-job npm package](https://www.npmjs.com/package/meteor-job), which
is fully interoperable with job-collection.

*NOTE!* Worker code very similar to what is shown below (without all of the DDP setup) can also run
on the Meteor server or even in a Meteor client. Everything before the call to `Job.processJobs()`
is just code to connect and authenticate a pure node.js script with a Meteor server.

```javascript
///////////////////
// node.js Worker
var DDP = require('ddp');
var DDPlogin = require('ddp-login');
var Job = require('meteor-job');

// `Job` here has essentially the same API as JobCollection on Meteor.
// In fact, job-collection is built on top of the 'meteor-job' npm package!

// Setup the DDP connection
var ddp = new DDP({
  host: "meteor.mydomain.com",
  port: 3000,
  use_ejson: true
});

// Connect Job with this DDP session
Job.setDDP(ddp);

// Open the DDP connection
ddp.connect(function (err) {
  if (err) throw err;
  // Call below will prompt for email/password if an
  // authToken isn't available in the process environment
  DDPlogin(ddp, function (err, token) {
    if (err) throw err;
    // We're in!
    // Create a worker to get sendMail jobs from 'myJobQueue'
    // This will keep running indefinitely, obtaining new work
    // from the server whenever it is available.
    // Note: If this worker was running within the Meteor environment,
    // Then only the call below is necessary to setup a worker!
    // However in that case processJobs is a method on the JobCollection
    // object, and not Job.
    var workers = Job.processJobs('myJobQueue', 'sendEmail',
      function (job, cb) {
        // This will only be called if a
        // 'sendEmail' job is obtained
        var email = job.data; // Only one email per job
        sendEmail(email.address, email.subject, email.message,
          function(err) {
            if (err) {
              job.log("Sending failed with error" + err,
                {level: 'warning'});
              job.fail("" + err);
            } else {
              job.done();
            }
            // Be sure to invoke the callback
            // when work on this job has finished
            cb();
          }
        );
      }
    );
  });
});
```

### Design

The design of job-collection is heavily influenced by [Kue](https://github.com/LearnBoost/kue) and
to a lesser extent by the [Maui Cluster Scheduler]
(https://en.wikipedia.org/wiki/Maui_Cluster_Scheduler). However, unlike Kue's use of Redis Pub/Sub
and an HTTP API, job-collection uses MongoDB, Meteor, and Meteor's DDP protocol to provide
persistence, reactivity, and secure remote access.

As the name implies, a `JobCollection` looks and acts like a Meteor Collection because under the
hood it actually is one. However, other than `.find()` and `.findOne()`, most accesses to a
`JobCollection` happen via the easy to use API on `Job` objects. Most `Job` API calls are
transformed internally to Meteor [Method](http://docs.meteor.com/#methods_header) calls. This is
cool because the underlying `Job` class is implemented as pure Javascript that can run in both the
Meteor server and client environments, and most significantly as pure node.js code running
independently from Meteor (as shown in the example code above).

## Installation

To add to your project, run:

    meteor add simonsimcity:job-collection

The package exposes a global object `JobCollection` on both client and server.

To run tests (using Meteor `meteortesting:mocha`) run from within the `packages/job-collection` subdirectory:

    meteor test-packages ./ --driver-package meteortesting:mocha

Or run the meteor-app this repository contains:

    npm run test

    # Logical OR was added because some tests currently fail.
    npm run coverage || npm run coverage:html

In case you ran the last line in your shell, you can find the coverage report in the `.coverage/lcov-report` folder.

A basic sample application that implements a basic image gallery with upload/download support and
automatic generation of thumbnail images is available. It also implements a basic job manager UI
that allows control of both individual jobs and changes to the entire collection at once. It is
available here: https://github.com/vsivsi/meteor-file-job-sample-app

## Use

job-collections are backed by [Meteor Collections](http://docs.meteor.com/#collections) and may be
used in similar ways. `.find()` and `.findOne()` work as you would expect and are fully reactive on
the client just as with a normal collection.

Other than the find methods mentioned above, interactions with a job collection occur using the
`JobCollection`, `Job` and `JobQueue` APIs documented below. The Job document model used in a job
collection is fully specified, maintained and enforced by the APIs.

Meteor clients are automatically denied permission to directly `insert`, `update` or `remove` jobs
from a job collection. To accomplish these types of tasks, a client must use the provided APIs,
subject to permissions set by specific allow/deny rules on the job collection. Servers retain access
to the standard `insert`, `update` or `remove` methods, but should avoid using them unless
absolutely necessary, favoring the job collection APIs to perform various tasks.

It is also possible (and highly useful!) to write your own clients outside of Meteor as vanilla
node.js programs using the `meteor-job` [npm package](https://www.npmjs.org/package/meteor-job),
which is actually used by job-collection internally, and implements essentially identical
functionality via the same interfaces.

### Security

Securing a job collection is done using mechanisms that will be familiar to anyone who has used the
[Meteor `publish` and `subscribe` mechanism](http://docs.meteor.com/#publishandsubscribe) and
[Meteor Collection allow/deny rules](http://docs.meteor.com/#allow).

For a client to have access to perform `find()` operations on a job collection, the server must
publish the collection and the client must subscribe to it. This works identically to normal Meteor
collections.

Compared to vanilla Meteor collections, job-collections have a  different set of remote methods with
specific security implications. Where the allow/deny methods on a Meteor collection take functions
to grant permission for `insert`, `update` and `remove`, job-collection has more functionality to
secure and configure.

There are currently over a dozen Meteor methods defined by each job-collection. In many cases it
will be most convenient to write allow/deny rules to one of the four predefined permission groups:
`admin`, `manager`, `creator` and `worker`. These defined roles separate security concerns and
permit you to efficiently add allow/deny rules for groups of functions that various client
functionalities are likely to need. Where these roles do not meet the requirements of a specific
project, each remote method can also be individually secured with custom allow/deny rules.

### Performance

The performance of job-collection will be almost entirely dependent on the speed of the MongoDB
server it is hosted on. By default job-collection creates these indexes in the underlying database:
```javascript
jc._ensureIndex({ type : 1, status : 1 });
jc._ensureIndex({ priority : 1, retryUntil : 1, after : 1 });
```
If you anticipate having large job collections (ie. with over 1000 jobs at a time) and you will be
doing custom queries on the database, you will want to create appropriate additional indexes to
ensure that your application performs well.

### Removing Old Jobs

Unless you do something to prevent it, completed and canceled jobs will accumulate in your database. In some scenarios this may be desirable, but if it's not, there are a few options to clean out old jobs from the database:

* Add a job cleaning job. This will take care of the most common use cases. It will allow you customize the logic to fit your specific needs. There's an example of a cleaning job in the "playground" [sample app](https://github.com/vsivsi/meteor-job-collection-playground/blob/576f46225d3dc3be81b1f89d6f7f57ad37789b12/play.coffee#L443-L464).
* Use [events](#jcevents---server-only) to remove jobs once they complete or are removed.

### Logging

The server can easily log all activity (both successes and failures) on a job collection by passing
any valid node.js writable Stream to `jc.setLogStream(writeStream)`. If you're just getting started try [`process.stdout`](https://nodejs.org/api/process.html#process_process_stdout) to log to your console.

Looking for more control over the output? Define a listener on the [jc.events](#jcevents---server) `call` event to implement custom logging.

## JobCollection API

### jc = new JobCollection([name], [options]) - *Anywhere*
#### Creates a new JobCollection

Creating a new `JobCollection` is similar to creating a new Meteor Collection. You simply specify a
name (which defaults to `"queue"`). On the server there are some additional methods you will
probably want to invoke on the returned object to configure it further.

`options`:

* `noCollectionSuffix` -- If `true`, `'.jobs'` won't be appended to the collection name. Default:
  `false`
* In addition, JobCollection supports the same options as Meteor [Mongo.Collection]
  (http://docs.meteor.com/#/full/mongo_collection)

For security and simplicity the traditional client allow/deny rules for Meteor collections are
preset to deny all direct client `insert`, `update` and `remove` type operations on a
`JobCollection`. This effectively channels all remote activity through the `JobCollection` DDP
methods, which may be secured using allow/deny rules specific to `JobCollection`. See the
documentation for `jc.allow()` and `jc.deny()` for more information.

```javascript
// the "new" is required
jc = new JobCollection('defaultJobCollection');
```

### jc.setLogStream(writeStream) - *Server only*
#### Sets where the job collection method invocation log will be written

You can log everything that happens to a job collection on the server by providing any valid
writable stream. You may only call this once, unless you first call `jc.shutdown()`, which will
automatically close the existing `logStream`.

```javascript
// Log everything to stdout
jc.setLogStream(process.stdout);
```

### jc.logConsole - *Client only*
#### Member variable that turns on DDP method call logging to the console

```javascript
jc.logConsole = false  // Default. Do not log method calls to the client console
```

### jc.promote([milliseconds]) - *Server only*
#### Sets time between checks for delayed jobs that are now ready to run

`jc.promote()` may be called at any time to change the polling rate. job-collection must poll for
this operation because it is time that is changing, not the contents of the database, so there are
no database updates to listen for.

```javascript
jc.promote(15*1000);  // Default: 15 seconds
```

Note: if you are running multiple Meteor instances that share access to a single job collection, you
can set the time each instance waits to promote to `N * milliseconds`, where `N` is the number of
Meteor instances. The instances will each take turns promoting jobs at 1/Nth of the desired rate.

### jc.allow(options) - *Server only*
#### Allow remote execution of specific job-collection methods

By default no remote operations are allowed, and in this configuration job-collection exists only as
a server-side service; with the creation, management and execution of all jobs dependent on
server-side Meteor code.


HISTORY.md
## Revision history

#### 1.6.1

* Replaced npm package `later` back by the equivalent atmosphere package after encounting major problems on the client side

#### 1.6.0

* Replaced meteor package `later` by the equivalent npm package
* Package updated to CoffeeScript 2. Now is required to create an instance of the class with `new JobCollection`, just calling `JobCollection` does not work anymore. Same for `new Job`.
* Replaced `tinytest` by `meteortesting:mocha` to run tests including code coverage

#### 1.5.2

* Fixed an issue under some circumstances when `job.cancel()` was run on a job with option `{ dependents: false }`. Thanks @gbhrdt.

#### 1.5.1

* Fixed issue introduced in 1.5.0 which caused `processJobs()` with arrays of jobType strings to fail a new type check.  Thanks @danielparas.

#### 1.5.0

* Added `errorCallback` option to `processJobs()`. This gives workers a way to log network and other errors and do something other than write them to `console.error` (the default and previous behavior).
* Added workaround for bug/feature in later.js that caused the first run of periodic jobs to sometimes occur immediately after scheduling. Thanks @mitar.
* Progress information is now preserved in the job document on 'done' or 'fail', previously it was set to hard coded values. Thanks @mitar.
* Removed section recommending capped collections for managing removal of old finished jobs. This no longer works as of MongoDB 3.2. Thanks @mcoenca.
* Jobs that complete before their dependents have even been saved will no longer result in unresolved dependencies. Thanks @serkandurusoy.
* Documentation improvements
* Updated package dependencies

### 1.4.0

* Added support for new "callbackStrict" option to `processJobs()`. When `true` (default `false`) `processJobs()` will throw an error if a worker function calls its callback more than once. Previously it only wrote a message to stderr in all cases. That functionality is preserved with this change.
* Added `job.repeatRetries` as an optional (for backward compatibility) attribute of the job document model. This attribute is used (when present) for the number of retries to attempt for the new instance of a repeating job. When missing (such as will be the case for jobs created by older versions of meteor-job) the old method of using `job.retries + job.retried` is used.
* Fix an error caused by a check on `retries` when a waiting/ready job is cancelled, restarted and then refreshed. This condition also occurs when `job.restart()` is called with option `retries` greater than the current value of `job.retried` (which is the default case after a cancel of a non-running job.) Thanks @huttarichard.
* In coordination with the above change, the `retries` option of `job.restart()` will now accept a value of zero.
* Fixed race condition in `processJobs()` that could cause `Job.getWork()` to run after the queue is paused or stopped, leading to a zombie job on the server.
* Documentation improvements
* Updated package dependencies

### 1.3.3

* Fixed error affecting all of the *Jobs functions which caused them to only process the first 256 jobs in the provided list of _id values.
* Updated package dependencies

### 1.3.2

* Fix broken links in README

### 1.3.1

* Updated package dependencies
* Documentation improvements (typos)

### 1.3.0

* Added `delayDeps` option to `job.done()` to allow dependent jobs to be delayed after the completion of an antecedent job. Thanks to @a4xrbj1 for suggesting.
* Fixed bug in `Job.jobStatuses`, Thanks @niceilm
* Invoking server-only or client-only methods in the wrong environment now generates a warning method. Thanks @aardvarkk.
* Additional indexes added to make built-in queries more efficient for large collections. Thanks @mitar.
* Internal improvements in event handlers. Thanks @mitar.
* Documentation improvements, Thanks @rhyslbw, @mitar and @jdcc !
* Updated job subproject to v1.3.2

### 1.2.3

* Fix issue when very long `promote()` cycles are used, causing waiting jobs that are ready to run at a server restart to be delayed one full promotion cycle. Thanks to @KrishnaPG for reporting this.

### 1.2.2

* Fixed bug in the `jc.getWork()` `workTimeout` functionality that could cause a running job to immediately auto-expire on the server if it was previously run with a `workTimeout` value, and was then subsequently run without a `workTimeout` value after the original job was rerun, restarted, retried or repeated.
* Updated job subproject to v1.3.1
* Updated README to include links to job-collection playground app and node.js worker repos

### 1.2.1

* Fixed `log()` on jobs without `runId` by loosening check in `jobLog()`. Thanks @sprohaska for the PR!

### 1.2.0

* Added ability for workers to specify a timeout for running jobs, so that if they crash or lose connectivity the job can automatically fail and be restarted. `getWork()` and `processJobs()` now each have a new option `workTimeout` that sets the number of milliseconds until a job can be automatically failed on the server. If not specified, the default behavior is as before (running jobs with no active worker need to be handled by the developer.) Thanks to @aldeed for the idea of making this a worker setting.
* Added `repeatId` option to `job.done()` which when `true` will cause the successful return value of a repeating job to be the `_id` of the newly scheduled job. Thanks to @tcastelli for this idea.
* Added `jc.events`, which is a node.js Event Emitter allowing server code to register callbacks to log or generate statistics based upon all job-collection DDP methods. There are two main events currently implemented: `'call'` for successful DDP method calls, and `'error'` for any errors thrown in such calls. There are also events defined for each of the 18 DDP methods (e.g. `jobDone` or `getWork`). These events emit for every such call, whether successful for not.

src/server.js
/**
 * Copyright (C) 2014-2017 by Vaughn Iverson
 * job-collection is free software released under the MIT/X11 license.
 * See included LICENSE file for details.
 */

import { Meteor } from "meteor/meteor";
// import Job from 'meteor-job';
import Job from "../job/src/job_class";
import { EventEmitter } from "events";
import JobCollectionBase from "./shared";

const userHelper = function(user, connection) {
  if (!connection) {
    return "[SERVER]";
  }
  return user || "[UNAUTHENTICATED]";
};

class JobCollection extends JobCollectionBase {
  constructor(root = "queue", options = {}) {
    // Call super's constructor
    super(root, options);

    this.events = new EventEmitter();

    this._errorListener = this.events.on("error", this._onError.bind(this));

    // Add events for all individual successful DDP methods
    this._methodErrorDispatch = this.events.on("error", msg =>
      this.events.emit(msg.method, msg)
    );

    this._callListener = this.events.on("call", this._onCall.bind(this));

    // Add events for all individual successful DDP methods
    this._methodEventDispatch = this.events.on("call", msg =>
      this.events.emit(msg.method, msg)
    );

    this.stopped = true;

    // No client mutators allowed
    super.deny.bind(this)({
      update: () => true,
      insert: () => true,
      remove: () => true
    });

    this.promote();

    this.logStream = null;

    this.allows = {};
    this.denys = {};

    // Initialize allow/deny lists for permission levels and ddp methods
    this.ddpPermissionLevels.concat(this.ddpMethods).forEach(level => {
      this.allows[level] = [];
      this.denys[level] = [];
    });

    // If a connection option is given, then this JobCollection is actually hosted
    // remotely, so don't establish local and remotely callable server methods in that case
    if (!options.connection) {
      // Default indexes, only when not remotely connected!
      this._ensureIndex({ type: 1, status: 1 });
      this._ensureIndex({ priority: 1, retryUntil: 1, after: 1 });
      this._ensureIndex({ priority: 1, retryUntil: 1, after: 1 });
      this._ensureIndex({ status: 1, after: 1 });
      this._ensureIndex({ status: 1, expiresAfter: 1 });

      this.isSimulation = false;
      const localMethods = this._generateMethods();
      this._localServerMethods = this._localServerMethods || {};

      Object.keys(localMethods).forEach(methodName => {
        const methodFunction = localMethods[methodName];
        this._localServerMethods[methodName] = methodFunction;
      });

      this._ddp_apply = (name, params, cb) => {
        if (cb) {
          return Meteor.defer(() => {
            let result;
            try {
              result = this._localServerMethods[name].apply(this, params);
            } catch (e) {
              return cb(e);
            }
            return cb(null, result);
          });
        } else {
          return this._localServerMethods[name].apply(this, params);
        }
      };

      Job._setDDPApply(this._ddp_apply, root);

      Meteor.methods(localMethods);
    }
  }

  _onError(msg) {
    const user = userHelper(msg.userId, msg.connection);
    return this._toLog(user, msg.method, `${msg.error}`);
  }

  _onCall(msg) {
    const user = userHelper(msg.userId, msg.connection);
    this._toLog(user, msg.method, `params: ${JSON.stringify(msg.params)}`);
    return this._toLog(
      user,
      msg.method,
      `returned: ${JSON.stringify(msg.returnVal)}`
    );
  }

  _toLog(userId, method, message) {
    return (
      this.logStream &&
      this.logStream.write(`${new Date()}, ${userId}, ${method}, ${message}\n`)
    );
  }

  _emit(method, connection, userId, err, ret, ...params) {
    if (err) {
      return this.events.emit("error", {
        error: err,
        method,
        connection,
        userId,
        params,
        returnVal: null
      });
    } else {
      return this.events.emit("call", {
        error: null,
        method,
        connection,
        userId,
        params,
        returnVal: ret
      });
    }
  }

  _methodWrapper(method, func) {
    const permitted = (userId, params) => {
      const performTest = tests =>
        tests.find(
          test =>
            (Array.isArray(test) && test.includes(userId)) ||
            (typeof test === "function" && test(userId, method, params))
        );

      const performAllTests = allTests =>
        this.ddpMethodPermissions[method].find(t => performTest(allTests[t]));

      return !performAllTests(this.denys) && performAllTests(this.allows);
    };
    // Return the wrapper function that the Meteor method will actually invoke
    return (...params) => {
      let retval;
      try {
        if (!this.connection || !!permitted(this.userId, params)) {
          retval = func(...params);
        } else {
          throw new Meteor.Error(
            403,
            "Method not authorized",
            "Authenticated user is not permitted to invoke this method."
          );
        }
      } catch (error) {
        this._emit(method, this.connection, this.userId, error);
        throw error;
      }
      this._emit(method, this.connection, this.userId, null, retval, ...params);
      return retval;
    };
  }

  setLogStream(writeStream) {
    if (this.logStream) {
      throw new Error(
        "logStream may only be set once per job-collection startup/shutdown cycle"
      );
    }

    this.logStream = writeStream;

    if (
      !this.logStream ||
      typeof this.logStream.write === "function" ||
      typeof this.logStream.end === "function"
    ) {
      throw new Error("logStream must be a valid writable node.js Stream");
    }
  }

  // Register application allow rules
  allow(allowOptions) {
    Object.keys(allowOptions)
      .filter(t => this.allows.hasOwnProperty(t))
      .forEach(t => this.allows[t].push(allowOptions[t]));
  }

  // Register application deny rules
  deny(denyOptions) {
    Object.keys(denyOptions)
      .filter(t => this.denys.hasOwnProperty(t))
      .forEach(t => this.denys[t].push(denyOptions[t]));
  }

  _DDPMethod_startJobServer(options = {}) {
    check(options, {});
    if (this.stopped && this.stopped !== true) {
      Meteor.clearTimeout(this.stopped);
    }
    this.stopped = false;
  }

  _DDPMethod_shutdownJobServer({ timeout = 60 * 1000 } = {}) {
    check(timeout, Match.Where(_validIntGTEOne));

    if (this.stopped && this.stopped !== true) {
      Meteor.clearTimeout(this.stopped);
    }

    this.stopped = Meteor.setTimeout(() => {
      const cursor = this.find(
        {
          status: status.RUNNING
        },
        {
          transform: null
        }
      );
      const failedJobs = cursor.count();
      if (failedJobs !== 0) {
        console.warn(`Failing ${failedJobs} jobs on queue stop.`);
      }
      cursor.forEach(d =>
        this._DDPMethod_jobFail(
          d._id,
          d.runId,
          "Running at Job Server shutdown."
        )
      );
      if (this.logStream != null) {
        // Shutting down closes the logStream!
        this.logStream.end();
        this.logStream = null;
        return this.logStream;
      }
    }, timeout);
  }

  _DDPMethod_getWork(type, { maxJobs = 1, workTimeout } = {}) {
    check(type, Match.OneOf(String, [String]));
    check(maxJobs, Match.Where(_validIntGTEOne));
    check(workTimeout, Match.Optional(Match.Where(_validIntGTEOne)));

    // Don't put out any more jobs while shutting down
    if (this.stopped) {
      return [];
    }

    // Support string types or arrays of string types
    if (typeof type === "string") {
      type = [type];
    }

    const time = new Date();
    let docs = [];
    const runId = this._makeNewID(); // This is meteor internal, but it will fail hard if it goes away.

    while (docs.length < maxJobs) {
      const ids = this.find(
        {
          type: {
            $in: type
          },
          status: status.READY,
          runId: null
        },
        {
          sort: {
            priority: 1,
            retryUntil: 1,
            after: 1
          },
          limit: maxJobs - docs.length, // never ask for more than is needed
          fields: {
            _id: 1
          },
          transform: null
        }
      ).map(d => d._id);

      if (!((ids != null ? ids.length : undefined) > 0)) {
        // Don't keep looping when there's no available work
        break;
      }

      const mods = {
        $set: {
          status: status.RUNNING,
          runId,
          updated: time
        },
        $inc: {
          retries: -1,
          retried: 1
        }
      };

      const logObj = JobCollectionBase._logMessage.running(runId);
      if (logObj) {
        mods.$push = { log: logObj };
      }

      if (typeof workTimeout === "number") {
        mods.$set.workTimeout = workTimeout;
        mods.$set.expiresAfter = new Date(time.valueOf() + workTimeout);
      } else {
        if (!mods.$unset) {
          mods.$unset = {};
        }
        mods.$unset.workTimeout = "";
        mods.$unset.expiresAfter = "";
      }

      const num = this.update(
        {
          _id: {
            $in: ids
          },
          status: status.READY,
          runId: null
        },
        mods,
        {
          multi: true
        }
      );

      if (num > 0) {
        let foundDocs = this.find(
          {
            _id: {
              $in: ids
            },
            runId
          },
          {
            fields: {
              log: 0,
              failures: 0,
              _private: 0
            },
            transform: null
          }
        ).fetch();

        if (foundDocs.length > 0) {
          if (this.scrub) {
            foundDocs = foundDocs.map(d => this.scrub(d));
          }
          check(docs, [_validJobDoc()]);
          docs = [...docs, ...foundDocs];
        }
      }
    }
    // else
    //   console.warn 'getWork: find after update failed'
    return docs;
  }


  _DDPMethod_jobReady(ids, { force = false, time = new Date() } = {}) {
    check(ids, Match.OneOf(Match.Where(_validId), [Match.Where(_validId)]));
    check(force, Boolean);
    check(time, Date);

    const now = new Date();

    if (_validId(ids)) {
      ids = [ids];
    }

    const query = {
      status: status.WAITING,
      after: {
        $lte: time
      }
    };

    const mods = {
      $set: {
        status: status.READY,
        updated: now
      }
    };

    if (ids.length > 0) {
      query._id = { $in: ids };
      mods.$set.after = now;
    }

    const logObj = [];

    if (force) {
      // Don't move to resolved, because they weren't!
      mods.$set.depends = [];
      const isForced = JobCollectionBase._logMessage.forced();
      if (isForced) {
        logObj.push(isForced);
      }
    } else {
      query.depends = { $size: 0 };
    }

    const readied = JobCollectionBase._logMessage.readied();
    if (readied) {
      logObj.push(readied);
    }

    if (logObj.length > 0) {
      mods.$push = {
        log: {
          $each: logObj
        }
      };
    }

    const num = this.update(query, mods, { multi: true });

    if (num > 0) {
      return true;
    } else {
      return false;
    }
  }

  // Hook function to sanitize documents before validating them in getWork() and getJob()
  scrub(job) {
    return job;
  }

  promote(milliseconds = 15 * 1000) {
    if (milliseconds == null) {
      milliseconds = 15 * 1000;
    }
    if (typeof milliseconds === "number" && milliseconds > 0) {
      if (this.interval) {
        Meteor.clearInterval(this.interval);
      }
      this._promote_jobs();
      this.interval = Meteor.setInterval(
        this._promote_jobs.bind(this),
        milliseconds
      );
      return this.interval;
    } else {
      return console.warn(
        `jobCollection.promote: invalid timeout: ${this.root}, ${milliseconds}`
      );
    }
  }

  _promote_jobs() {
    if (this.stopped) {
      return;
    }

    // This looks for zombie running jobs and autofails them
    this.find({ status: "running", expiresAfter: { $lt: new Date() } }).forEach(
      job =>
        new Job(this.root, job).fail(
          "Failed for exceeding worker set workTimeout"
        )
    );

    // Change jobs from waiting to ready when their time has come
    // and dependencies have been satisfied
    return this.readyJobs();
  }

  // Warning Stubs for client-only calls
  logConsole() {
    throw new Error("Client-only function jc.logConsole() invoked on server.");
  }
}

export default JobCollection;
